lib/componentManager.ts

```diff
@@ -0,0 +1,608 @@
+import {
+  ComponentAction,
+  ContentType,
+  Environment,
+  SNItem
+} from "@standardnotes/snjs";
+import { environmentToString, generateUuid, isValidJsonString } from "./utils";
+import Logger from "./logger";
+
+const DEFAULT_COALLESED_SAVING_DELAY = 250;
+
+enum MessagePayloadApi {
+  Component = "component",
+}
+
+type Component = {
+  uuid?: string;
+  origin?: string;
+  data?: Record<string, any>;
+  sessionKey?: string;
+  environment?: string;
+  platform?: string;
+  isMobile?: boolean;
+  acceptsThemes?: boolean;
+  activeThemes?: string[];
+}
+
+type MessagePayload = {
+  action: ComponentAction;
+  data: Record<string, any>;
+  componentData?: Record<string, any>;
+  messageId?: string;
+  sessionKey?: string;
+  api: MessagePayloadApi;
+  original?: MessagePayload;
+  callback?: (...params: any) => void;
+}
+
+type ComponentManagerOptions = {
+  coallesedSaving?: boolean,
+  coallesedSavingDelay?: number,
+  debug?: boolean,
+  acceptsThemes?: boolean
+}
+
+type PermissionObject = {
+  name: ComponentAction,
+  content_types?: ContentType[]
+}
+
+type ComponentManagerParams = {
+  initialPermissions?: PermissionObject[]
+  options?: ComponentManagerOptions,
+  onReady?: () => void
+}
+
+type ItemPayload = {
+  content_type?: ContentType,
+  content?: Record<string, any>,
+  [key: string]: any
+}
+
+export default class ComponentManager {
+  private initialPermissions?: PermissionObject[];
+  private onReadyCallback?: () => void;
+  private component: Component = { activeThemes: [], acceptsThemes: true };
+  private sentMessages: MessagePayload[] = [];
+  private messageQueue: MessagePayload[] = [];
+  private lastStreamedItem?: SNItem;
+  private pendingSaveItems?: SNItem[];
+  private pendingSaveTimeout?: NodeJS.Timeout;
+  private pendingSaveParams?: any;
+  private coallesedSaving = false;
+  private coallesedSavingDelay = DEFAULT_COALLESED_SAVING_DELAY;
+  private messageHandler?: (event: any) => void;
+
+  constructor(private contentWindow: Window, params?: ComponentManagerParams) {
+    if (!contentWindow) {
+      throw new Error("contentWindow must be a valid Window object.");
+    }
+    if (params) {
+      this.processParameters(params);
+    }
+    this.registerMessageHandler();
+  }
+
+  private processParameters(params: ComponentManagerParams) {
+    const { initialPermissions, options, onReady } = params;
+
+    if (initialPermissions && initialPermissions.length > 0) {
+      this.initialPermissions = initialPermissions;
+    }
+    if (options?.coallesedSaving) {
+      this.coallesedSaving = options.coallesedSaving;
+    }
+    if (options?.coallesedSavingDelay) {
+      this.coallesedSavingDelay = options.coallesedSavingDelay;
+    }
+    if (options?.acceptsThemes) {
+      this.component.acceptsThemes = options?.acceptsThemes ?? true;
+    }
+    if (onReady) {
+      this.onReadyCallback = onReady;
+    }
+    Logger.enabled = options?.debug ?? false;
+  }
+
+  public deinit() {
```

Comment:
all methods should declare return type. That's the bigest advantage of using TypeScript. 

You can ensure that with the `eslint` setup I proposed as an example.